File: dlox/lib/token.dart

enum TokenType {
  // Single-character tokens.
  leftParen, rightParen, leftBrace, rightBrace,
  comma, dot, minus, plus, semicolon, slash, star,

  // one or two character tokens.
  bang, bangEqual,
  equal, equalEqual,
  greater, greaterEqual,
  less, lessEqual,

  // literals.
  identifier, string, decimal, integer,

  // keywords.
  and, classType, elseType, falseType, func, forType, ifType, nil, or,
  print, returnType, superType, thisType, trueType, varType, whileType,

  eof, comment
}

class Token {
  final TokenType type;
  final String lexeme;
  final dynamic literal;
  final int line;

  Token(this.type, this.lexeme, this.literal, this.line);

  @override
  String toString() {
    return "Token($type, $lexeme, $line)";
  }
}

File: src/token.zig
const LineInfo = @import("./lineInfo.zig").LineInfo;
pub const Token = struct {
    @"type": TokenType,
    lexeme: []const u8,
    line: LineInfo,

    pub fn blank() Token {
        return Token{ .@"type" = TokenType.@"error", .lexeme = &[_]u8{}, .line = LineInfo{
            .line = 0,
            .column = 0,
        } };
    }
};

pub const TokenType = enum(u8) {
    // Single-character tokens.
    left_paren,
    right_paren,
    left_brace,
    right_brace,
    comma,
    dot,
    minus,
    plus,
    semicolon,
    slash,
    star,
    // one or two character tokens.
    bang,
    bang_equal,
    equal,
    equal_equal,
    greater,
    greater_equal,
    less,
    less_equal,
    // literals.
    identifier,
    string,
    integer,
    float,
    // keywords.
    @"and",
    class,
    @"else",
    @"false",
    @"for",
    fun,
    @"if",
    nil,
    @"or",
    print,
    @"return",
    super,
    this,
    @"true",
    @"var",
    @"while",

    @"error",
    eof,
};
